jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -0,0 +1,407 @@
+                    loss_actor = -jnp.minimum(loss_actor1, loss_actor2)
+                    loss_actor = loss_actor.mean()
+                    entropy = actor_policy.entropy().mean()
+
+                    total_loss = (
+                        loss_actor
+                        + config["VF_COEF"] * value_loss
+                        - config["ENT_COEF"] * entropy
+                    )
+                    return total_loss, (value_loss, loss_actor, entropy)
+
+                grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+                total_loss, grads = grad_fn(
+                    params, opt_state, traj_batch, advantages, targets
+                )
+
+                grads, total_loss = jax.lax.pmean(
+                    (grads, total_loss), axis_name="batch"
+                )
+                grads, total_loss = jax.lax.pmean(
+                    (grads, total_loss), axis_name="device"
+                )
+
+                updates, new_opt_state = update_fn(grads, opt_state)
+                new_params = optax.apply_updates(params, updates)
+
+                return (new_params, new_opt_state), total_loss
+
+            params, opt_state, traj_batch, advantages, targets, rng = update_state
+            rng, _rng = jax.random.split(rng)
+
+            # SHUFFLE MINIBATCHES
+            batch_size = config["ROLLOUT_LENGTH"] * config["NUM_ENVS"]
+            permutation = jax.random.permutation(_rng, batch_size)
+            batch = (traj_batch, advantages, targets)
+            batch = jax.tree_util.tree_map(
+                lambda x: x.reshape((batch_size,) + x.shape[2:]), batch
+            )
+            shuffled_batch = jax.tree_util.tree_map(
+                lambda x: jnp.take(x, permutation, axis=0), batch
+            )
+            minibatches = jax.tree_util.tree_map(
+                lambda x: jnp.reshape(
+                    x, [config["NUM_MINIBATCHES"], -1] + list(x.shape[1:])
+                ),
+                shuffled_batch,
+            )
+
+            # UPDATE MINIBATCHES
+            (params, opt_state), total_loss = jax.lax.scan(
+                _update_minibatch, (params, opt_state), minibatches
+            )
+
+            update_state = (params, opt_state, traj_batch, advantages, targets, rng)
+            return update_state, total_loss
+
+        update_state = (params, opt_state, traj_batch, advantages, targets, rng)
+
+        # UPDATE EPOCHS
+        update_state, loss_info = jax.lax.scan(
+            _update_epoch, update_state, None, config["PPO_EPOCHS"]
+        )
+
+        params, opt_state, traj_batch, advantages, targets, rng = update_state
+        runner_state = (params, opt_state, rng, env_state, last_timestep)
+        metric = traj_batch.info
+        return runner_state, metric
+
+    def learner_fn(
+        params: FrozenDict,
+        opt_state: OptState,
+        rng: chex.PRNGKey,
+        env_state: State,
+        timesteps: TimeStep,
+    ) -> dict:
```

Comment:
If you implement runner state as mentioned above.

```suggestion
    ) -> Dict[str, Union[chex.Array, RunnerState]]:
```

Reply:
Even better if this function returns a `NamedTuple` instead of a dict, since you're only ever accessing the outputs, you're never adding more fields to the dict